pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -61,12 +62,141 @@ type replicaCircuitBreaker struct {
 	stopper *stop.Stopper
 	r       replicaInCircuitBreaker
 	st      *cluster.Settings
+	cancels CancelStorage
 	wrapped *circuit.Breaker
+
+	versionIsActive int32 // atomic
+}
+
+// Register takes a cancelable context and its cancel function, and registers
+// them with the circuit breaker. If the breaker is already tripped, its error
+// is returned immediately and the caller should not continue processing the
+// request. Otherwise, the cancel function is invoked when the breaker trips.
+// The caller is provided with a token and signaller for use in a call to
+// Unregister upon request completion. That method also takes the error (if any)
+// resulting from the request to ensure that in the case of a tripped breaker,
+// the error reflects this fact.
+func (br *replicaCircuitBreaker) Register(
+	ctx context.Context, cancel func(),
+) (_token interface{}, _ signaller, _ error) {
+	brSig := br.Signal()
+
+	if brSig.C() == nil {
+		// Circuit breakers are disabled, so don't do any work registering the
+		// context. Unregister will know that we didn't since it checks the
+		// same brSig.
+		return ctx, brSig, nil
+	}
+
+	// TODO(tbg): we may want to exclude more requests from this check, or allow
+	// requests to exclude themselves from the check (via their header). This
+	// latter mechanism could also replace isCircuitBreakerProbe.
+	if isCircuitBreakerProbe(ctx) {
+		brSig = neverTripSignaller{}
```

Comment:
Can't we move this above the `brSig.C() == nil` check above to also exclude these from tracking?

Reply:
Actually, wouldn't these leak if we don't, since the `C() == nil` check in `Unregister` would prevent them from being removed?

Reply:
Good catch! You're right, requests from the probe currently enter the registry but are never removed. This was not intentional. I'm switching the order which should fix this, but leaving a TODO to add test coverage of this case.